### DISCUSSION TITLE
Fuzz testing with google/oss-fuzz

### DISCUSSION THREAD

**Discussion by nathaniel-brough** [typst/typst#2388] [Ideas]
Hey Typst Team,

I hope this message finds you well. I've been following along with Typst for some time now, and I really like the direction that it's heading. I'd like to suggest and champion an effort to set up some basic fuzz-testing and combine it with google/oss-fuzz for continuous fuzzing. I'm fully aware that you are very busy people and I don't want to overload your review/maintenance capacity by introducing too many new ideas. Is this a bad time to discuss potential security/reliability improvements? 

If your not familiar with fuzzing or oss-fuzz I've included a few brief notes below.

#### **Benefits of Fuzz-Testing**

- **Dynamic Code Testing**: Fuzz-testing challenges systems with unexpected data, aiming to identify vulnerabilities. It’s akin to an exhaustive stress-test for the code.
- **Detecting Hidden Vulnerabilities**: It can uncover potential weaknesses that may not be evident in routine tests.
- **Continuous and Automated Testing**: With tools like Google’s OSS-Fuzz, fuzz-testing can be automated, running continuously on distributed systems, ensuring daily resilience checks.

#### **Google/oss-fuzz for Continuous Fuzzing**

- **Automated Fuzzing**: OSS-Fuzz undertakes comprehensive fuzz-testing daily on a distributed cluster.
- **Security Boost**: It provides enhanced security measures free of cost, thanks to Google’s backing.
- **Detailed Reporting**: OSS-Fuzz offers exhaustive reports in case of detected anomalies, enabling effective action.

To further emphasize the significance of security and best practices, [OpenSSF](https://openssf.org/best-practices/) offers a wealth of information on the matter.

I’d be more than happy to lead the effort in integrating fuzz testing with Typst and assist in any way required.

As a proof of concept I created a super simple minimal fuzz harness for the `typst-syntax` crate in  #2581.

**Comment by laurmaedje** [comment]
Thanks for opening up the discussion and the PR. I'm especially busy right now because of the upcoming release, but I will get back to this once Typst 0.9 is out.
